src/liballoc/btree/map.rs

```diff
@@ -890,6 +892,11 @@ impl<K: Ord, V> BTreeMap<K, V> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn entry(&mut self, key: K) -> Entry<K, V> {
+        // FIXME(@porglezomp) Avoid allocating if we don't insert
+        if self.root.is_shared_root() {
+            self.root = node::Root::new_leaf();
+        }
```

Comment:
Could you abstract this pattern into a method? `ensure_root_is_owned()` or something?